xmtp_mls/src/groups/mod.rs

```diff
@@ -459,7 +461,8 @@ impl MlsGroup {
         let update_interval = Some(5_000_000);
         self.maybe_update_installations(conn.clone(), update_interval, client)
             .await?;
-        self.publish_intents(conn, client).await?;
+        self.publish_intents(conn.clone(), client).await?;
+        self.sync_until_all_intents_resolved(conn, client).await?;
```

Comment:
This is a lot of overhead to handle a pretty fringe-y edge case. Will have impacts on latency (now we have to wait for at least 2 sequential API requests to complete before returning on every message send) and on rate-limiting.

Given that most clients are going to receive the message from a stream anyways I'm not sure it's necessary.

Reply:
The last paragraph of the PR description has a justification for this - WDYT?

> Will have impacts on latency 
Not to perceived latency if optimistic message sends are used. If non-optimistic sends are used, i.e. confirmation of message send is desired, I'm not sure if it's honest to return success until we have synced.

> on rate-limiting
I think it'll be negligible, as we have significantly smaller publish volume compared to reads

> Given that most clients are going to receive the message from a stream anyways I'm not sure it's necessary.
Receiving it from a stream won't cause the client to retry

Reply:
> Not to perceived latency if optimistic message sends are used. If non-optimistic sends are used, i.e. confirmation of message send is desired, I'm not sure if it's honest to return success until we have synced.

For an inbox app, separating prepare and send makes sense. But this is also going to effect bulk senders/bots who are almost certainly not using optimistic sends.

> Receiving it from a stream won't cause the client to retry

That's fair. Updated my comment.


Maybe the better question is: if we are going to pay the price for adding a read to every write, couldn't we just sync first and prevent this problem in almost all cases? It would take a hell of a race condition to get 3 commits in between the sync and the send.

Reply:
Also keep in mind that perceived latency on the receiver side is not affected - the publish is going to take the same amount of time regardless

Reply:
With the next version of the protocol, there's an opportunity to have this validation done server-side (reject the message if a more recent commit has been seen), perhaps we could remove this extra round-trip when that happens

Reply:
Sorry didn't see your last comment until just now! Funny given we're talking about stale local state.

>  this is also going to effect bulk senders/bots who are almost certainly not using optimistic sends

Are bulk senders sending multiple messages back-to-back within the same conversation, or single messages to many conversations? Is it the case that they need to be up-to-date with their groups regardless?

>  if we are going to pay the price for adding a read to every write, couldn't we just sync first and prevent this problem in almost all cases? It would take a hell of a race condition to get 3 commits in between the sync and the send

We could. I'm not sure if it's that much of a race condition in large groups, and it seems we've already run into it in our own testing. Also, integrating apps are not guaranteed to have a subscription running, and we're not guaranteed that a running subscription isn't broken or has a network delay etc.

The problem is even if we accept that it's rare, we have no recovery mechanism otherwise. Those messages are permanently lost. We could talk about a different way to recover, but it might involve a large refactor. I'm inclined to wait until our work on the new decentralized backend ships, when we will have some degree of server-side validation.

Reply:
> it's not clear that getting client apps to sync every time they open a thread, or on some interval, will create a lower overall volume than syncing every time a send happens.

It might be the same volume, but I think that has to happen anyways to catch missing messages. So this would be purely additional